client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala

```diff
@@ -1,97 +1,54 @@
-    .followRedirects(true)
-    .connectionPool(new ConnectionPool(10, 60, TimeUnit.SECONDS))
-    .build()
-
-  val targetUrl = "https://content.guardianapis.com"
-
-  def item(id: String) = ItemQuery(id)
-  val search = SearchQuery()
-  val tags = TagsQuery()
-  val sections = SectionsQuery()
-  val editions = EditionsQuery()
-  val removedContent = RemovedContentQuery()
-  val atoms = AtomsQuery()
-  val recipes = RecipesQuery()
-  val reviews = ReviewsQuery()
-  val gameReviews = GameReviewsQuery()
-  val restaurantReviews = RestaurantReviewsQuery()
-  val filmReviews = FilmReviewsQuery()
-  val videoStats = VideoStatsQuery()
-  val stories = StoriesQuery()
-
-  case class HttpResponse(body: Array[Byte], statusCode: Int, statusMessage: String)
-
-  protected[client] def url(location: String, parameters: Map[String, String]): String = {
-    require(!location.contains('?'), "must not specify parameters in URL")
-
-    location + QueryStringParams(parameters + ("api-key" -> apiKey) + ("format" -> "thrift"))
-  }
-
-  protected def fetch(url: String)(implicit context: ExecutionContext): Future[Array[Byte]] = {
-    val headers = Map("User-Agent" -> userAgent, "Accept" -> "application/x-thrift")
-
-    for (response <- get(url, headers)) yield {
-      if (List(200, 302) contains response.statusCode) response.body
-      else throw contentApiError(response)
-    }
-  }
-
-  private def contentApiError(response: HttpResponse): GuardianContentApiError = {
-    val errorResponse = Try(ThriftDeserializer.deserialize(response.body, ErrorResponse)).toOption
-    GuardianContentApiError(response.statusCode, response.statusMessage, errorResponse)
-  }
-
-  protected def get(url: String, headers: Map[String, String])(implicit context: ExecutionContext): Future[HttpResponse] = {
-
-    val reqBuilder = new Request.Builder().url(url)
-    val req = headers.foldLeft(reqBuilder) {
-      case (r, (name, value)) => r.header(name, value)
-    }
-
-    val promise = Promise[HttpResponse]()
-
-    http.newCall(req.build()).enqueue(new Callback() {
-      override def onFailure(call: Call, e: IOException): Unit = promise.failure(e)
-      override def onResponse(call: Call, response: Response): Unit = {
-        promise.success(HttpResponse(response.body().bytes, response.code(), response.message()))
-      }
-    })
-
-    promise.future
-  }
-
-  def getUrl(contentApiQuery: ContentApiQuery): String =
-    url(s"$targetUrl/${contentApiQuery.pathSegment}", contentApiQuery.parameters)
-
+import com.gu.contentapi.client.utils.QueryStringParams
+import scala.concurrent.{ExecutionContext, Future}
+
+trait ContentApiClient {
+  
+  /** Your API key */
+  def apiKey: String
+  
+  /** The user-agent identifier */
+  def userAgent: String
+  
+  /** The url of the CAPI endpoint */
+  def targetUrl: String
+
+  /** Queries CAPI.
+    *
+    * This method must make a GET request to the CAPI endpoint
+    * and streamline the response into an HttpResponse object.
```

Comment:
I think that this paragraph, combined with the method signature, highlights the improvements that we need in this PR.

The method signature is generic enough, accepting a URL and HTTP headers and expecting a response. It's generic enough so that there is no need to mention CAPI at all. Indeed, the method can be moved to a separate trait, and its docs would just say that should fulfil any HTTP(S) call.

Reply:
Yeah I thought about that, essentially writing a custom interpreter would be
```scala
object MyClientAPI extends ContentApiLogic with MyHttpAPI
```
It feels clean enough to me for now. Ultimately, my intent is to provide an interpreter of the CAPI algebra into a separate REST algebra so that there is nothing in CAPI suggesting a particular communication protocol.